core/commands/patch.go

```diff
@@ -0,0 +1,132 @@
+package commands
+
+import (
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	mh "github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/go-multihash"
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+
+	cmds "github.com/ipfs/go-ipfs/commands"
+	u "github.com/ipfs/go-ipfs/util"
+)
+
+var PatchCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Create a new merkledag object based on an existing one",
+		ShortDescription: `
+'ipfs patch <root> [add-link|rm-link] <args>' is a plumbing command used to
+build custom DAG objects.
+
+Examples:
+
+	EMPTY_DIR=QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn 
+	BAR=$(echo "bar" | ipfs add -q)
+    ipfs patch $EMPTY_DIR add-link foo $BAR
+
+This takes an empty directory, and adds a link named foo under it, pointing to
+a file containing 'bar', and returns the hash of the new object.
+
+	ipfs patch $FOO_BAR rm-link foo
+
+This removes the link named foo from the hash in $FOO_BAR and returns the
+resulting object hash.
+`,
+	},
+	Options: []cmds.Option{},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("root", true, false, "the hash of the node to modify"),
+		cmds.StringArg("command", true, false, "the operation to perform"),
+		cmds.StringArg("args", true, true, "extra arguments"),
+	},
+	Type: u.Key(""),
+	Run: func(req cmds.Request, res cmds.Response) {
+		nd, err := req.Context().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		rhash := u.B58KeyDecode(req.Arguments()[0])
+		if rhash == "" {
+			res.SetError(fmt.Errorf("incorrectly formatted root hash"), cmds.ErrNormal)
+			return
+		}
+
+		ctx, _ := context.WithTimeout(req.Context().Context, time.Second*30)
+		rnode, err := nd.DAG.Get(ctx, rhash)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		switch req.Arguments()[1] {
+		case "add-link":
```

Comment:
As I [explain here][1], I'd also like an `add-or-replace-link` action (can you think of a more compact name?) to avoid either (a) adding two links with the same name or (b) requiring separate calls with `rm-link` and `add-link`.  Not that (b) is so bad, but I expect this to be a common activity, so adding some sugar to handle it seems to be worth the cost of a separate action.

[1]: https://github.com/ipfs/go-ipfs/pull/1187#issuecomment-106696791